### DISCUSSION TITLE
Unable to use package in tenancy mode

### DISCUSSION THREAD

**Discussion by lucabecchetti** [spatie/laravel-settings#257] [Q&A]
hi guys... i'm using this wonderfull package with spatie settings, but it always try to load settings from central datasse, i create a new model class that extend spatie settings model:

```php
<?php

namespace App\Models;
use Spatie\LaravelSettings\Models\SettingsProperty as BaseSettingsProperty;
use Stancl\Tenancy\Database\Concerns\TenantConnection;

class TenantSettingsProperty extends BaseSettingsProperty
{
    use TenantConnection;
}

```
But i get: Database connection [tenant] not configured

Can someone help me please?

**Comment by rubenvanassche** [comment]
Converted to a discussion since nothing is broken in the package. I myself don't use Tenancy, so no idea how to solve this. I think there's another discussion here somewhere with some hints to an implementation.

**Comment by nidhalkratos** [comment]
I have literally spent two days to find a solution for this, if you find my answer helpful feel free to buy me a coffee at PayPal using this address iknowbrain@gmail.com.
Alright, the solution is to create two different repositories, one for the central connection and another for the tenant connections.

<img width="756" alt="image" src="https://github.com/spatie/laravel-settings/assets/10298337/351454f0-835a-453c-96ed-5688685e101f">

And then within your settings class, you can override the method repository() to return correct repository accordingly.
<img width="396" alt="image" src="https://github.com/spatie/laravel-settings/assets/10298337/ebf2f529-e94e-4e2a-93f6-10117cb00653">

For example in my case, I created a `CentralSettings`  trait with the repository method defined, and then I use the trait in the settings class which needs to connect to the central database.

Of course, don't forget to change the `default_repository` in the settings config file to be 'database_tenant`

**Comment by farisfaikar** [comment]
I'm having a weird problem when using the `spatie/laravel-settings` package and `tenancy-for-laravel` package where in the `settings` table, for every column `group` that has the value of `general`, the tenant side reads from the central database instead of the tenant database. I'm using multi-database, so this shouldn't be the case. The weird part is, that tenancy works just fine on every other table and even in the same `settings` table that's in a different group. What's even weirder is that the data is updated just fine in the tenant database. It's just the read operation that's messed up and it's reading from central instead of tenant. I've reviewed the codes and I can't find any difference between the other setting classes.

I'm wondering if you guys are using `tenancy-for-laravel` package for your tenant resolution package the issue might be related to that package as well.

**Comment by nick-potts** [comment]
Thought I'd just throw this in here because it works perfectly for me.

I use a custom repository:

```php
class SettingsRepository extends DatabaseSettingsRepository
{
    public function __construct(array $config)
    {
        parent::__construct($config);
        Assert::notNull(tenant());
    }

    public function updatePropertiesPayload(string $group, array $properties): void
    {
        $propertiesInBatch = collect($properties)->map(function ($payload, $name) use ($group) {
            return [
                'group' => $group,
                'name' => $name,
                'payload' => $this->encode($payload),
                'tenant_id' => tenant()->getTenantKey()
            ];
        })->values()->toArray();

        $this->getBuilder()
            ->where('group', $group)
            ->upsert($propertiesInBatch, ['group', 'name', 'tenant_id'], ['payload']);
    }
}
```

Then I modified the settings table migration to have a tenant_id:

```php
Schema::create('settings', function (Blueprint $table): void {
    $table->id();

    $table->string('tenant_id');

    $table->string('group');
    $table->string('name');
    $table->boolean('locked')->default(false);
    $table->json('payload');

    $table->timestamps();

    $table->unique(['group', 'name', 'tenant_id']);
});
```

Then a custom model:
```php
class SpatieSettings extends SettingsProperty
{
    use BelongsToTenant;
}
```

Then when you create a setting migration, place it in the tenant migration folder and run `php artisan tenants:migrate`

With this setup, you can just use the package like normal and it'll pull per tenant.
